device/common/src/seeding/make_doublet_spM_buffers.cpp

```diff
@@ -0,0 +1,70 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2022 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Library include(s).
+#include "traccc/seeding/device/make_doublet_spM_buffers.hpp"
+
+// VecMem include(s).
+#include <vecmem/containers/vector.hpp>
+
+// System include(s).
+#include <algorithm>
+#include <cassert>
+
+namespace traccc::device {
+
+doublet_spM_buffer_pair make_doublet_spM_buffers(
+    const doublet_counter_container_types::const_view& doublet_counter,
+    vecmem::copy& copy, vecmem::memory_resource& mr,
+    vecmem::memory_resource* mr_host) {
+
+    // Get the number of doublets per geometric bin.
+    vecmem::vector<device::doublet_counter_header> doublet_counts(
+        (mr_host != nullptr) ? mr_host : &mr);
+    copy(doublet_counter.headers, doublet_counts);
+    vecmem::jagged_vector<device::doublet_counter> doublet_counts_items(
+        (mr_host != nullptr) ? mr_host : &mr);
+    copy(doublet_counter.items, doublet_counts_items);
```

Comment:
This is the expensive bit. Copying the jagged vector back to the host is expensive here.

You can speed things up by specifying explicitly that you're doing a `vecmem::copy::type::device_to_host` copy. But we'll need to discuss how the logic here should happen. As doing this copy is just a bad idea in the first place... :frowning:

Reply:
Yeah, I'm aware this is the major issue with the current implementation, but I can't quite figure out a way to make it work without this

Reply:
Just for the record as I will mark this as resolved, I took Attila's suggestion of changing the the whole data structure of not only the previous kernel (count doublets) as well as the downstream kernels to spM based jagged vectors. I will now do some performance testing and cleaning up of code to check wether it is worth merging this in.